File: bindings/ocaml/lib/proven_smtp.ml
(* SPDX-License-Identifier: MPL-2.0 *)
(* Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk> *)

(** SMTP protocol bindings for proven-servers.

    Wraps the C-ABI functions from
    [protocols/proven-smtp/ffi/zig/src/smtp.zig]. *)

(** SMTP session states. *)
type session_state =
  | Connected        (** TCP connection established. *)
  | Greeted          (** HELO/EHLO completed. *)
  | Auth_started     (** AUTH exchange in progress. *)
  | Authenticated    (** Successfully authenticated. *)
  | Mail_from        (** MAIL FROM accepted. *)
  | Rcpt_to          (** RCPT TO accepted. *)
  | Data             (** DATA transfer in progress. *)
  | Message_received (** End-of-data received. *)
  | Quit             (** Session ended. *)

(** SMTP AUTH mechanisms. *)
type auth_mechanism =
  | Plain      (** PLAIN mechanism. *)
  | Login      (** LOGIN mechanism. *)
  | Cram_md5   (** CRAM-MD5 mechanism. *)
  | Xoauth2    (** XOAUTH2 mechanism. *)

let state_to_tag = function
  | Connected -> 0 | Greeted -> 1 | Auth_started -> 2 | Authenticated -> 3
  | Mail_from -> 4 | Rcpt_to -> 5 | Data -> 6 | Message_received -> 7
  | Quit -> 8

let state_of_tag = function
  | 0 -> Some Connected | 1 -> Some Greeted | 2 -> Some Auth_started
  | 3 -> Some Authenticated | 4 -> Some Mail_from | 5 -> Some Rcpt_to
  | 6 -> Some Data | 7 -> Some Message_received | 8 -> Some Quit
  | _ -> None

let auth_mech_to_tag = function
  | Plain -> 0 | Login -> 1 | Cram_md5 -> 2 | Xoauth2 -> 3

let auth_mech_of_tag = function
  | 0 -> Some Plain | 1 -> Some Login | 2 -> Some Cram_md5
  | 3 -> Some Xoauth2 | _ -> None

(* --- C FFI declarations --- *)

external c_smtp_abi_version : unit -> int = "smtp_abi_version"
external c_smtp_create_context : unit -> int = "smtp_create_context"
external c_smtp_destroy_context : int -> unit = "smtp_destroy_context"
external c_smtp_get_state : int -> int = "smtp_get_state"
external c_smtp_get_reply_code : int -> int = "smtp_get_reply_code"
external c_smtp_get_recipient_count : int -> int = "smtp_get_recipient_count"
external c_smtp_get_data_size : int -> int = "smtp_get_data_size"
external c_smtp_get_auth_mechanism : int -> int = "smtp_get_auth_mechanism"
external c_smtp_is_authenticated : int -> int = "smtp_is_authenticated"
external c_smtp_is_tls_active : int -> int = "smtp_is_tls_active"
external c_smtp_greet : int -> int -> int = "smtp_greet"
external c_smtp_authenticate : int -> int -> int = "smtp_authenticate"
external c_smtp_auth_complete : int -> int -> int = "smtp_auth_complete"
external c_smtp_set_sender : int -> int = "smtp_set_sender"
external c_smtp_add_recipient : int -> int = "smtp_add_recipient"
external c_smtp_start_data : int -> int = "smtp_start_data"
external c_smtp_append_data : int -> int -> int = "smtp_append_data"
external c_smtp_finish_data : int -> int = "smtp_finish_data"
external c_smtp_reset : int -> int = "smtp_reset"
external c_smtp_quit : int -> int = "smtp_quit"
external c_smtp_enable_tls : int -> int = "smtp_enable_tls"
external c_smtp_can_transition : int -> int -> int = "smtp_can_transition"

(* --- Safe wrappers --- *)

let abi_version () = c_smtp_abi_version ()
let create_context () = Proven_error.from_slot (c_smtp_create_context ())
let destroy_context slot = c_smtp_destroy_context slot
let get_state slot = state_of_tag (c_smtp_get_state slot)
let get_reply_code slot = c_smtp_get_reply_code slot
let get_recipient_count slot = c_smtp_get_recipient_count slot
let get_data_size slot = c_smtp_get_data_size slot
let get_auth_mechanism slot = auth_mech_of_tag (c_smtp_get_auth_mechanism slot)
let is_authenticated slot = c_smtp_is_authenticated slot = 1
let is_tls_active slot = c_smtp_is_tls_active slot = 1

let greet slot ~ehlo =
  Proven_error.from_status (c_smtp_greet slot (if ehlo then 1 else 0))

let authenticate slot mech =
  Proven_error.from_status (c_smtp_authenticate slot (auth_mech_to_tag mech))

let auth_complete slot ~success =
  Proven_error.from_status (c_smtp_auth_complete slot (if success then 1 else 0))

let set_sender slot = Proven_error.from_status (c_smtp_set_sender slot)
let add_recipient slot = Proven_error.from_status (c_smtp_add_recipient slot)
let start_data slot = Proven_error.from_status (c_smtp_start_data slot)
let append_data slot ~len = Proven_error.from_status (c_smtp_append_data slot len)
let finish_data slot = Proven_error.from_status (c_smtp_finish_data slot)
let reset slot = Proven_error.from_status (c_smtp_reset slot)
let quit slot = Proven_error.from_status (c_smtp_quit slot)
let enable_tls slot = Proven_error.from_status (c_smtp_enable_tls slot)

let can_transition ~from ~to_ =
  c_smtp_can_transition (state_to_tag from) (state_to_tag to_) = 1

File: bindings/php/src/ProvenSmtp.php
<?php

// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// PHP bindings for the proven-smtp Zig FFI.

declare(strict_types=1);

namespace ProvenServers;

/** SMTP session states matching Idris2 ABI tags. */
enum SmtpSessionState: int
{
    case Connected       = 0;
    case Greeted         = 1;
    case AuthStarted     = 2;
    case Authenticated   = 3;
    case MailFrom        = 4;
    case RcptTo          = 5;
    case Data            = 6;
    case MessageReceived = 7;
    case Quit            = 8;
}

/** SMTP AUTH mechanisms matching Idris2 ABI tags. */
enum SmtpAuthMechanism: int
{
    case Plain   = 0;
    case Login   = 1;
    case CramMd5 = 2;
    case Xoauth2 = 3;
}

/**
 * SMTP session context wrapping a Zig FFI slot.
 */
final class ProvenSmtp
{
    private const CDEF = <<<'CDEF'
    int smtp_create_context(void);
    void smtp_destroy_context(int slot);
    uint8_t smtp_get_state(int slot);
    uint16_t smtp_get_reply_code(int slot);
    uint32_t smtp_get_recipient_count(int slot);
    uint32_t smtp_get_data_size(int slot);
    uint8_t smtp_get_auth_mechanism(int slot);
    uint8_t smtp_is_authenticated(int slot);
    uint8_t smtp_is_tls_active(int slot);
    uint8_t smtp_greet(int slot, uint8_t ehlo);
    uint8_t smtp_authenticate(int slot, uint8_t mechanism);
    uint8_t smtp_auth_complete(int slot, uint8_t success);
    uint8_t smtp_set_sender(int slot);
    uint8_t smtp_add_recipient(int slot);
    uint8_t smtp_start_data(int slot);
    uint8_t smtp_append_data(int slot, uint32_t length);
    uint8_t smtp_finish_data(int slot);
    uint8_t smtp_reset(int slot);
    uint8_t smtp_quit(int slot);
    uint8_t smtp_enable_tls(int slot);
    uint32_t smtp_abi_version(void);
    uint8_t smtp_can_transition(uint8_t from, uint8_t to);
    CDEF;

    private static ?\FFI $ffi = null;
    private int $slot;
    private bool $destroyed = false;

    private function __construct(int $slot) { $this->slot = $slot; }

    private static function ffi(): \FFI
    {
        if (self::$ffi === null) {
            self::$ffi = ProvenServers::loadLibrary('smtp', self::CDEF);
        }
        return self::$ffi;
    }

    /** @throws ProvenError */
    public static function create(): self
    {
        return new self(ProvenError::checkSlot(self::ffi()->smtp_create_context()));
    }

    public function destroy(): void
    {
        if (!$this->destroyed) {
            self::ffi()->smtp_destroy_context($this->slot);
            $this->destroyed = true;
        }
    }

    public function getState(): ?SmtpSessionState
    {
        $tag = self::ffi()->smtp_get_state($this->slot);
        return $tag <= 8 ? SmtpSessionState::from($tag) : null;
    }

    public function getReplyCode(): int { return self::ffi()->smtp_get_reply_code($this->slot); }
    public function getRecipientCount(): int { return self::ffi()->smtp_get_recipient_count($this->slot); }
    public function getDataSize(): int { return self::ffi()->smtp_get_data_size($this->slot); }

    public function getAuthMechanism(): ?SmtpAuthMechanism
    {
        $tag = self::ffi()->smtp_get_auth_mechanism($this->slot);
        return $tag <= 3 ? SmtpAuthMechanism::from($tag) : null;
    }

    public function isAuthenticated(): bool { return self::ffi()->smtp_is_authenticated($this->slot) === 1; }
    public function isTlsActive(): bool { return self::ffi()->smtp_is_tls_active($this->slot) === 1; }

    /** @throws ProvenError */
    public function greet(bool $ehlo = true): void
    {
        ProvenError::checkStatus(self::ffi()->smtp_greet($this->slot, $ehlo ? 1 : 0));
    }

    /** @throws ProvenError */
    public function authenticate(SmtpAuthMechanism $mechanism): void
    {
        ProvenError::checkStatus(self::ffi()->smtp_authenticate($this->slot, $mechanism->value));
    }

    /** @throws ProvenError */
    public function authComplete(bool $success): void
    {
        ProvenError::checkStatus(self::ffi()->smtp_auth_complete($this->slot, $success ? 1 : 0));
    }

    /** @throws ProvenError */
    public function setSender(): void { ProvenError::checkStatus(self::ffi()->smtp_set_sender($this->slot)); }
    /** @throws ProvenError */
    public function addRecipient(): void { ProvenError::checkStatus(self::ffi()->smtp_add_recipient($this->slot)); }
    /** @throws ProvenError */
    public function startData(): void { ProvenError::checkStatus(self::ffi()->smtp_start_data($this->slot)); }

    /** @throws ProvenError */
    public function appendData(int $length): void
    {
        ProvenError::checkStatus(self::ffi()->smtp_append_data($this->slot, $length));
    }

    /** @throws ProvenError */
    public function finishData(): void { ProvenError::checkStatus(self::ffi()->smtp_finish_data($this->slot)); }
    /** @throws ProvenError */
    public function reset(): void { ProvenError::checkStatus(self::ffi()->smtp_reset($this->slot)); }
    /** @throws ProvenError */
    public function quit(): void { ProvenError::checkStatus(self::ffi()->smtp_quit($this->slot)); }
    /** @throws ProvenError */
    public function enableTls(): void { ProvenError::checkStatus(self::ffi()->smtp_enable_tls($this->slot)); }

    public static function abiVersion(): int { return self::ffi()->smtp_abi_version(); }
    public static function canTransition(SmtpSessionState $from, SmtpSessionState $to): bool
    {
        return self::ffi()->smtp_can_transition($from->value, $to->value) === 1;
    }
}
